### PR TITLE
Adds fields to specification for initialize the disassembly context in a function 

### PR DESCRIPTION
**Pull request by 2over12**

### CODE REVIEW

**Comment by 2over12** [comment]
Blocking on https://github.com/lifting-bits/remill/pull/622

**Comment by 2over12** [comment]
There are a fair few cases where we will label an interprocedural branch as in  intraprocedural branch in thumb and arm, this will be addressed by the more control flow spec so I do not think we need to address this here

File: include/anvill/Specification.h [hunk 1]
```diff
@@ -64,7 +65,10 @@ struct ControlFlowTargetList final {
   // is, a given indirect jump may target the same address in multiple different
   // ways (e.g. multiple `case` labels in a `switch` statement that share the
   // same body).
-  std::set<std::uint64_t> target_addresses;
+
+  /// The addresses map a target to a given context mapping.
+  std::map<std::uint64_t, std::map<std::string, uint64_t>> target_addresses;
```
Comment (pgoodman) [inline include/anvill/Specification.h:70]: Is ordering in the second significant, as compared to the prior unordered_map?

File: lib/Lifters/FunctionLifter.cpp [hunk 2]
```diff
@@ -322,17 +326,30 @@ void FunctionLifter::VisitNoOp(
...
   VisitDelayedInstruction(inst, delayed_inst, block, true);
   llvm::BranchInst::Create(
-      GetOrCreateTargetBlock(inst, inst.branch_taken_pc, mapper), block);
+      GetOrCreateTargetBlock(inst, mapper.taken_flow.known_target,
+                             mapper.taken_flow.static_context),
+      block);
+}
+
+remill::DecodingContext FunctionLifter::ApplyTargetList(
+    const std::map<std::string, uint64_t> assignments,
```
Comment (pgoodman) [inline lib/Lifters/FunctionLifter.cpp:339]: `&assignments`

Comment (pgoodman) [reply lib/Lifters/FunctionLifter.cpp:339]: or drop the `const`.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: include/anvill/Specification.h (hunk 1 above)
Comment (2over12) [reply include/anvill/Specification.h]: Updated it can be unordered

File: Dockerfile [hunk 3]
```diff
@@ -17,30 +17,31 @@ ADD https://github.com/lifting-bits/remill/releases/latest/download/remill_ubunt
...
+ARG CXX_COMMON_VERSION
+ARG DEBIAN_FRONTEND=noninteractive 
 RUN apt-get update && \
-    apt-get install -qqy --no-install-recommends libdbus-1-3 curl unzip python3 python3-pip python3.8 python3.8-venv python3-setuptools xz-utils && \
+    apt-get install -qqy --no-install-recommends git libdbus-1-3 curl unzip python3 python3-pip python3.8 python3.8-venv python3-setuptools xz-utils cmake && \
     rm -rf /var/lib/apt/lists/*
 
 WORKDIR /dependencies
 
+ADD https://github.com/lifting-bits/cxx-common/releases/download/v${CXX_COMMON_VERSION}/vcpkg_ubuntu-${UBUNTU_VERSION}_llvm-${LLVM_VERSION}_amd64.tar.xz vcpkg_ubuntu-${UBUNTU_VERSION}_llvm-${LLVM_VERSION}_amd64.tar.xz
```
Comment (Ninja3047) [inline Dockerfile:36]: i believe cxx-common is already preset in this build step's base image (FROM trailofbits/cxx-common-vcpkg-builder-ubuntu:${UBUNTU_VERSION} as deps)

**Comment by frabert** [comment]
Could you document the changes made to the spec format in the md file?

Also, reading through this made me realize I have no idea what our policy re: `LOG` vs `DLOG` is

**Comment by frabert** [comment]
Could you document the changes made to the spec format in the md file?

Also, reading through this made me realize I have no idea what our policy re: `LOG` vs `DLOG` is

File: bin/Decompile/Main.cpp [hunk 4]
```diff
@@ -229,9 +229,13 @@ int main(int argc, char *argv[]) {
 
   spec.ForEachFunction([&lifter, &names, &target_funcs](auto decl) {
     llvm::Function *func;
-    if (target_funcs.empty() ||
+    if (FLAGS_lift_list.empty() ||
         target_funcs.find(decl->address) != target_funcs.end()) {
+      LOG(INFO) << "attempting to lift: " << std::hex << decl->address;
```
Comment (frabert) [inline bin/Decompile/Main.cpp:234]: `DLOG`

File: bin/Decompile/Main.cpp [hunk 5]
```diff
@@ -229,9 +229,13 @@ int main(int argc, char *argv[]) {
...
       func = lifter.LiftEntity(*decl);
+      if (!func) {
+        LOG(ERROR) << "Failed to lift: " << std::hex << decl->address;
```
Comment (frabert) [inline bin/Decompile/Main.cpp:237]: `DLOG`

File: lib/Lifters/FunctionLifter.cpp [hunk 6]
```diff
@@ -271,6 +273,7 @@ FunctionLifter::DecodeInstructionInto(const uint64_t addr, bool is_delayed,
     return options.arch->DecodeDelayedInstruction(
         addr, inst_out->bytes, *inst_out, std::move(context));
   } else {
+    LOG(INFO) << "Ops emplace: " << inst_out->operands.size();
```
Comment (frabert) [inline lib/Lifters/FunctionLifter.cpp:276]: `DLOG`

File: lib/Lifters/FunctionLifter.cpp [hunk 7]
```diff
@@ -1677,7 +1508,7 @@ llvm::Function *FunctionLifter::DeclareFunction(const FunctionDecl &decl) {
 // Lift a function. Will return `nullptr` if the memory is
 // not accessible or executable.
 llvm::Function *FunctionLifter::LiftFunction(const FunctionDecl &decl) {
-
+  LOG(INFO) << "Attempting to lift inside function lifter";
```
Comment (frabert) [inline lib/Lifters/FunctionLifter.cpp:1511]: `DLOG`

File: lib/Lifters/FunctionLifter.cpp [hunk 8]
```diff
@@ -1765,13 +1599,18 @@ llvm::Function *FunctionLifter::LiftFunction(const FunctionDecl &decl) {
...
+  for (const auto &[k, v] : decl.context_assignments) {
+    default_mapping.UpdateContextReg(k, v);
+  }
+
   ir.CreateBr(GetOrCreateBlock(0u, func_address, default_mapping));
 
   AnnotateInstructions(entry_block, pc_annotation_id,
                        GetPCAnnotation(func_address));
 
+  LOG(INFO) << "Visiting insns";
```
Comment (frabert) [inline lib/Lifters/FunctionLifter.cpp:1613]: `DLOG`

**Review by frabert — CHANGES_REQUESTED** [review]
I think this is everything I feel comfortable providing feedback on!

File: bin/Decompile/Main.cpp (hunk 5 above)
Comment (2over12) [reply bin/Decompile/Main.cpp:237]: Since this is a binary not a lib, logging the error is probably ideal

**Review by frabert — APPROVED** [review]
LGTM!

**Pull request merged**